### ISSUE TITLE
`nixery.dev` style images with non-sequential layers break `buildah bud`

### ISSUE THREAD

**Issue by SohumB** [containers/buildah#4025]
**Description**

```shell
$ cat Dockerfile
FROM nixery.dev/shell
RUN ls
$ buildah bud .
STEP 1/2: FROM nixery.dev/shell
STEP 2/2: RUN ls
bin
dev
etc
lib
libexec
nix
nix-support
proc
run
share
sys
tmp
usr
COMMIT
error committing container for step {Env:[SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[ls] Flags:[] Attrs:map[] Message:RUN ls Original:RUN ls}: error copying layers and metadata for container "b2dfabd70db97640d06962855cf9bef4cf618c883cf005b495c2e86904792f67": initializing source containers-storage:shell-working-container: internal error: history lists 1 non-empty layers, but we have 10 layers on disk
```

I believe this is how `nixery.dev` works: the layers it packages together in one manifest do not follow the standard sequential pattern. That seems to be what's confusing buildah here.

```shell
$ apt list buildah
Listing... Done
buildah/experimental,now 1.26.1+ds1-1 amd64 [installed]
buildah/testing,unstable 1.23.1+ds1-2+b1 i386
$ buildah version
Version:         1.26.1
Go Version:      go1.18.2
Image Spec:      1.0.2-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        1.0.0
libcni Version:  
image Version:   5.21.1
Git Commit:      
Built:           Thu Jan  1 10:00:00 1970
OS/Arch:         linux/amd64
BuildPlatform:   linux/amd64
$ cat /etc/*release
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
$ uname -a
Linux dynast 5.17.0-3-amd64 #1 SMP PREEMPT Debian 5.17.11-1 (2022-05-26) x86_64 GNU/Linux
$ cat ~/.config/containers/storage.conf
[storage]
driver = "overlay"
runroot = "/run/user/1000"
graphroot = "/home/sohum/.local/share/containers/storage"
[storage.options]
mount_program = "/usr/bin/fuse-overlayfs"
```

**Comment by SohumB** [comment]
Absolutely!

Additionally, there appear to be environment issues under these images too. I'm not sure if this is this bug or a different bug; let me know if I should open a new issue for this. 

```shell
$ cat Dockerfile
FROM nixery.dev/shell
RUN env
$ docker build . 
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM nixery.dev/shell
 ---> e0dc353c8e2d
Step 2/2 : RUN env
 ---> Running in 3473c9fd7497
HOSTNAME=3473c9fd7497
PWD=/
HOME=/
SHLVL=0
SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env
Removing intermediate container 3473c9fd7497
 ---> b5ca3705948c
Successfully built b5ca3705948c
$ buildah bud .
STEP 1/2: FROM nixery.dev/shell
STEP 2/2: RUN env
HOSTNAME=6871a14d19d6
PWD=/
SHLVL=0
SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env
COMMIT
error committing container for step {Env:[SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[env] Flags:[] Attrs:map[] Message:RUN env Original:RUN env}: error copying layers and metadata for container "6871a14d19d63249910dc5f0d4e5727009ef794535f81cd131ef44cae8f61e9a": initializing source containers-storage:shell-working-container: internal error: history lists 1 non-empty layers, but we have 10 layers on disk
$ buildah bud --env HOME=/ .
STEP 1/2: FROM nixery.dev/shell
STEP 2/2: RUN env
HOSTNAME=4be8aabea01c
PWD=/
SHLVL=0
SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env
COMMIT
error committing container for step {Env:[SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[env] Flags:[] Attrs:map[] Message:RUN env Original:RUN env}: error copying layers and metadata for container "4be8aabea01c25b3be8d9c2e5af2a9e351fe4585d54d2bb97242928ea071efed": initializing source containers-storage:shell-working-container: internal error: history lists 1 non-empty layers, but we have 10 layers on disk
```

**Comment by flouthoc** [comment]
Hi @SohumB, Would you be able to tell us from which build tool  `nixery.dev/shell` was created cause it contains 10 layers but `History` in spec marks only one. This usually does not happens I am just wondering which build tool was used. But anyways since `history` is optional in `image-spec` so I think buildah should drop this sanity check https://github.com/opencontainers/image-spec/blob/main/config.md#properties so following diff should fix this for you.

```diff
diff --git a/image.go b/image.go
index e3668bd0..3c24e709 100644
--- a/image.go
+++ b/image.go
@@ -558,17 +558,6 @@ func (i *containerImageRef) NewImageSource(ctx context.Context, sc *types.System
        dimage.History = append(dimage.History, dnews)
        appendHistory(i.postEmptyLayers)
 
-       // Sanity check that we didn't just create a mismatch between non-empty layers in the
-       // history and the number of diffIDs.
-       expectedDiffIDs := expectedOCIDiffIDs(oimage)
-       if len(oimage.RootFS.DiffIDs) != expectedDiffIDs {
-               return nil, errors.Errorf("internal error: history lists %d non-empty layers, but we have %d layers on disk", expectedDiffIDs, len(oimage.RootFS.DiffIDs))
-       }
-       expectedDiffIDs = expectedDockerDiffIDs(dimage)
-       if len(dimage.RootFS.DiffIDs) != expectedDiffIDs {
-               return nil, errors.Errorf("internal error: history lists %d non-empty layers, but we have %d layers on disk", expectedDiffIDs, len(dimage.RootFS.DiffIDs))
-       }
-
        // Encode the image configuration blob.
        oconfig, err := json.Marshal(&oimage)
        if err != nil {
```

But changes has associated issue in the past so just getting ack from @nalind on this.

**Comment by SohumB** [comment]
I don't precisely know the details of how `nixery.dev` works, but the short version is that it's an experiment in building images out of independent layers to maximise cache efficiency. When you request, say, `nixery.dev/git/htop`, the server uses the `nix` package manager to generate `git` and `htop`'s independent filesystems, and then just puts them together in a single OCI image.

There's some documentation on this process [here](https://nixery.dev/nixery.html) and [here](https://grahamc.com/blog/nix-and-layered-docker-images) and [here](https://storage.googleapis.com/nixdoc/nixery-layers.html). The source code is [here](https://github.com/tazjin/nixery).

**Comment by flouthoc** [comment]
PR https://github.com/containers/buildah/pull/4028 should close use-cases like this with `buildah build --add-history=false .`

**Comment by vrothberg** [comment]
Note that Docker happily builds the upper Dockerfile.  If users have to do an extra step, Buildah would break compat.

**Comment by flouthoc** [comment]
> Note that Docker happily builds the upper Dockerfile. If users have to do an extra step, Buildah would break compat.

@vrothberg Ah i see although docker does not looks like its doing correct thing cause it appends new history even when base has a `history:null` but in-order to maintain parity :) added another commit for docker-parity which still adds history even if baseImage history is `null` and ignore sanity check so `buildah build .` should also work for this issue once this is merged https://github.com/containers/buildah/pull/4028 but still recommend way would be `--add-history=false`.
